### PR TITLE
AURORA4 recipe

### PR DESCRIPTION
**Pull request by 27jiangziyan**
Very coarse result on AURORA4 dataset.

### CODE REVIEW

**Comment by ShigekiKarita** [comment]
Do you have any results on the eval subsets;  A, B, C, D like the existing papers? 
https://ieeexplore.ieee.org/document/7552554
> The evaluation set is derived from the WSJ0 5K-word closed vocabulary test set, which consists of 330 utterances from 8 speakers. This test set was recorded by the primary microphone and one secondary microphone. Each of these two parts is then corrupted by the same six noises used in the training set at 5-15 dB SNR, creating a total of 14 test sets. Notice that the types of noise are common across the training and test sets but SNRs of the data are not. These 14 test sets can then be grouped into 4 subsets: clean, noisy, clean with channel distortion, and noisy with channel distortion, which will be referred to as A, B, C, and D, respectively.

**Comment by 27jiangziyan** [comment]
I didn't group it into 4 subsets. The RESULTS file shows the result on the whole dataset.(14*330 utterances in total.) I follow the WSJ recipe and adjust several parameters. The result isn't good right now. I'm now working to add the LM and tune the parameters in the training process.
BTW, do you think I should get the results seperately as is shown in the existing papers?

**Comment by sw005320** [comment]
@27jiangziyan, I summarized your action items

- [ ] use config files (see the other v.0.4.0 recipes)
- [ ] use word-based RNNLM (see the WSJ recipe)
- [ ] mix the original training data with WSJ SI284 clean data (see chime4 or reverb recipes)
- [ ] arrange the RESULTS file to follow the Aurora-4 standard way (as @ShigekiKarita pointed out)
- [ ] use wide-shallow network (encoder: 3-layer 1024 units, decoder: 1-layer 1024 units)

Please check the performance step by step.

**Comment by sw005320** [comment]
@27jiangziyan, please do not push your local changes, e.g., `cmd_backend` in `cmd.sh` and `stage`, `stop_stage`, `ngpu` in `run.sh` and others. Also, do not include your local changes of the voxforge recipe.

**Comment by 27jiangziyan** [comment]
Finish all the action items only without arranging the RESULTS.(will do it finally.) The RESULTS report now only use noisy data.(not use clean data from train_si284.But the recipe include that.)

**Comment by sw005320** [comment]
Could you remove your local changes in the wsj and voxforge recipes?

File: egs/aurora4/asr1/local/wsj_data_prep.sh [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+   cat ${x}_sph.scp | awk '{print $1}' | perl -ane 'chop; m:^...:; print "$_ $&\n";' > $x.utt2spk
+   cat $x.utt2spk | $utils/utt2spk_to_spk2utt.pl > $x.spk2utt || exit 1;
+done
+
+
+#in case we want to limit lm's on most frequent words, copy lm training word frequency list
+cp links/13-32.1/wsj1/doc/lng_modl/vocab/wfl_64.lst $lmdir
+chmod u+w $lmdir/*.lst # had weird permissions on source.
+
+# The 20K vocab, open-vocabulary language model (i.e. the one with UNK), without
```
Comment (sw005320) [inline egs/aurora4/asr1/local/wsj_data_prep.sh]: Remove the LM related lines below from here to line 184

File: egs/aurora4/asr1/local/aurora4_data_prep.sh [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+utils=`pwd`/utils
+
+. ./path.sh # Needed for KALDI_ROOT
+sph2pipe=$KALDI_ROOT/tools/sph2pipe_v2.5/sph2pipe
+if [ ! -x $sph2pipe ]; then
+  echo "Could not find (or execute) the sph2pipe program at $sph2pipe";
+  exit 1;
+fi
+
+if [ -z $IRSTLM ] ; then
```
Comment (sw005320) [inline egs/aurora4/asr1/local/aurora4_data_prep.sh]: Remove IRSTLM related part from here to line 42.
We don't need it.

File: egs/aurora4/asr1/local/wsj_data_prep.sh [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+utils=`pwd`/utils
+
+. ./path.sh # Needed for KALDI_ROOT
+sph2pipe=$KALDI_ROOT/tools/sph2pipe_v2.5/sph2pipe
+if [ ! -x $sph2pipe ]; then
+  echo "Could not find (or execute) the sph2pipe program at $sph2pipe";
+  exit 1;
+fi
+
+if [ -z $IRSTLM ] ; then
```
Comment (sw005320) [inline egs/aurora4/asr1/local/wsj_data_prep.sh]: Remove IRSTLM related part from here to line 37.
We don't need it.

**Comment by 27jiangziyan** [comment]
I have removed my local changes in other recipes. Sorry for my error.

File: egs/aurora4/asr1/local/aurora4_data_prep.sh [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+done
+
+# Make the utt2spk and spk2utt files.
+for x in train_si84_clean train_si84_multi test_eval92 test_0166 dev_0330 dev_1206; do
+  cat ${x}_sph.scp | awk '{print $1}' \
+    | perl -ane 'chop; m:^...:; print "$_ $&\n";' > $x.utt2spk
+  cat $x.utt2spk | $utils/utt2spk_to_spk2utt.pl > $x.spk2utt || exit 1;
+done
+
+#in case we want to limit lm's on most frequent words, copy lm training word frequency list
```
Comment (sw005320) [inline egs/aurora4/asr1/local/aurora4_data_prep.sh]: Remove the LM related lines below from here to line 382

File: egs/aurora4/asr1/local/wsj_prepare_dict.sh [hunk 5]
```diff
@@ -0,0 +1,83 @@
+#!/bin/bash
```
Comment (sw005320) [inline egs/aurora4/asr1/local/wsj_prepare_dict.sh]: I believe we don't need this (we don't need a dictionary).
Please check it.

File: egs/aurora4/asr1/run.sh [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+. ./path.sh
+. ./cmd.sh
+
+# general configuration
+backend=pytorch
+stage=0        # start from 0 if you need to start from data preparation
+stop_stage=5
```
Comment (sw005320) [inline egs/aurora4/asr1/run.sh]: use the original number (100?)

File: egs/aurora4/asr1/run.sh [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+# decoding parameter
+recog_model=model.acc.best # set a model to be used for decoding: 'model.acc.best' or 'model.loss.best'
+
+# data
+aurora4=/export/corpora5/AURORA
+wsj0=/export/corpora5/LDC/LDC93S6B
+wsj1=/export/corpora5/LDC/LDC94S13B
+
+# exp tag
+tag="wide_shallow_network" # tag for managing experiments.
```
Comment (sw005320) [inline egs/aurora4/asr1/run.sh]: `tag=""`. 
Again, please do not put your local setup.

File: egs/aurora4/asr1/run.sh [hunk 8]
```diff
@@ -0,0 +1,284 @@
+#!/bin/bash
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
```
Comment (sw005320) [inline egs/aurora4/asr1/run.sh]: Change this to your name.

File: egs/aurora4/asr1/run.sh [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+    fbankdir=fbank
+    # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
+    # for x in train_si84_multi dev_0330 test_0166 train_si284; do
+    for x in train_si84_multi dev_0330 test_0166; do
+        steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 32 --write_utt2num_frames true \
+            data/${x} exp/make_fbank/${x} ${fbankdir}
+        utils/fix_data_dir.sh data/${x}
+    done
+
+    # echo "combine train_si84_multi and train_si284"
```
Comment (sw005320) [inline egs/aurora4/asr1/run.sh]: Remove comment outed lines.

File: egs/aurora4/asr1/local/wsj_data_prep.sh (hunk 1 above)
Comment (27jiangziyan) [reply egs/aurora4/asr1/local/wsj_data_prep.sh]: Maybe we could also remove the LM related lines in WSJ recipe?

Comment (sw005320) [reply egs/aurora4/asr1/local/wsj_data_prep.sh]: that is a great idea. we can do it later.

**Comment by 27jiangziyan** [comment]
Have reflected your reviews.
Have finished experiment by mixing si284 training data and put it in the RESULTS.
One remaining thing is to split the recog set into four subsets and show it in the RESULTS. I think I could do it by changing score_sclite.sh a little bit. But I think it may be complicated to include it in my recipe(for example, just use a command pipeline.) So could I just show the result?

**Comment by ShigekiKarita** [comment]
Maybe you can check swbd+fisher task. It shows CERs of swbd, fisher and callhome respectively
https://github.com/espnet/espnet/blob/ae79281050b60db25a2860ea7b0fa54470b1e64a/egs/fisher_swbd/asr1/local/score_sclite.sh

**Comment by 27jiangziyan** [comment]
Thank for your advice. I think it could work.

**Comment by sw005320** [comment]
Yes, we should have some specific local scoring scripts for this.
Also, it is quite informative if you keep your old results (first result -> wide-shallow -> word RNNLM -> mix SI284 training data) instead of replacing results.

**Comment by 27jiangziyan** [comment]
Add local scoring script and organize RESULTS as you suggest.

File: egs/aurora4/asr1/run.sh [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+. ./path.sh
+. ./cmd.sh
+
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+train_set=train_mix
```
Comment (sw005320) [inline egs/aurora4/asr1/run.sh]: Could you please add some comments about `train_mix` and the original `train_si84_multi` (this is the AURORA4 standard training data, right?) here?
Then, the user could select `train_mix` based custom training set or `train_si84_multi` based original set to report their numbers.

Comment (27jiangziyan) [reply egs/aurora4/asr1/run.sh]: Yes, I have added comments about the training dataset.

**Comment by sw005320** [comment]
Great job @27jiangziyan!
